### DISCUSSION TITLE
How IF( ) works ?

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#412] [Q&A] [answered]
Really broke my head trying to understand the users manual on how the function IF works in this cone 3D example. :)))

 number_density = if((r gt ri) and (r lt ro), den_cone, 0.0)
 number_density = if((x gt xi) and (x lt xo) and (r lt ri), den_cone, number_density(proton))
 number_density = if(x gt xo, 0.0, number_density(proton))

I do not find the definition of IF. Is it possible to get at least one complete figure with exact coordinates? There exist one in the manual but on it it is unclear if the density is constant inside the cone material. I understand the  laser entrance is at x=-10um in this example. 

Let's just start from the last simplest equation

 number_density = if(x gt xo, 0.0, number_density(proton))

Why there are two numbers here - the 0.0 and number_density(proton)? To set the slope? What is initial point for the slope and where is final? If this is not a slope and the density is constant when x gt xo then why there are two numbers instead of just one number_density(proton)? Again if this is not a slope but the abrupt jump - was density set to  0 before x0 and became number_density(proton) after x0 ? 

Same with the first IF. If the intention is to set the number_density of the cone material equal to den_cone what for another number 0.0 is there which if there is no definition how IF works (you have to remember that) is possible to interpret multiple ways instead of just one ?

The second IF in the setup above i even afraid to touch being completely totally confused ...

**Comment by Status-Mirror** [comment, marked as answer]
Hey Dan,

The if() function is explained in the maths parser section:
https://epochpic.github.io/documentation/code_details/maths_parser.html

This example also makes heavy use of the constant block:
https://epochpic.github.io/documentation/input_deck/input_deck_constant.html

In epoch3d, each cell has an x, y and z co-ordinate, corresponding to the cell-centre position. These number density lines aim to set the number density value in each cell. In an earlier constants block, the parameters xi, xo are set by the user (see the constant block). Here we also define r as shorthand for sqrt(y^2 + z^2), where y and z refer to the y and z co-ordinates of each cell. Finally, we define functions for ri and ro, which vary with the x parameter. These set the inner and outer radii of the cone, along with the slope.

The if(a,b,c) function is fed 3 arguments. The first is a logical test applied to every cell in the domain, the second is a value if that test is true, the third is a value if that test is false.

The first line reads: _number_density = if((r gt ri) and (r lt ro), den_cone, 0.0)_. The logical test applied to every cell is to see if r is between ri and ro. For cells with x, y and z which satisfy these criteria, the number density is set to den_cone, another parameter set by the user in the constant block. The other cells have their number density set to 0.

Now we have defined a number density for this species - we can actually use this number density in future lines. The second line reads: _number_density = if((x gt xi) and (x lt xo) and (r lt ri), den_cone, number_density(proton))_. This says that for cells between the xo, xi limits, and with r < ri (inside the cone), the number density should be den_cone. The remaining cells are set to number_density(proton), which refers to the previous number density of this species (i.e., don't change the number density in these cells).

The final line is much the same - for cells with x over xo, set their number density to 0. The remaining cells keep the number-density they had before this line was read.

The net result here is between x_min and xi, the cone is hollow. From xi to xo, the cone is filled, and beyond xo there is no cone. By stacking IF statements together like this, along with definining functions in constant blocks, the user can build complex geometries. This is just a particularly complicated example of using EPOCH if statements.

**Comment by DanRRRR** [comment]
Thanks Stuart. So it works like that 

if(a,b,c) - Conditional function. If a != 0 the function returns b, otherwise the function returns c.

The problem solved... 

But I need to digest why it was designed like this... So this actually looks like an analog of IF -THEN - ELSE.   
Thanks for the help again
